notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb

```diff
@@ -0,0 +1,628 @@
+    "        df_dict[metric_name] = [metric_func(preds) if use_preds else metric_func(scores) \n",
+    "                                for model_name, (preds, scores) in models_dict.items()]\n",
+    "    return pd.DataFrame.from_dict(df_dict, orient=\"index\", columns=models_dict.keys())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We calculate several performance and disparity metrics below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Metrics\n",
+    "models_dict = {\"Unmitigated\": (test_preds, test_scores)}\n",
+    "get_metrics_df(models_dict, Y_test, A_str_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As the overall performance metric we will use the _area under ROC curve_ (AUC), which is suited to classification problems with a large imbalance of positive and negative examples. For binary classifiers, this is the same as _balanced accuracy_.\n",
+    "\n",
+    "As the fairness metric we will use *equalized odds difference*, which quantifies the disparity in accuracy experienced by different demographics. Our goal is to assure that neither of the two groups (men vs women) has substantially larger false-positive rates or false-negative rates than the other group. The equalized odds difference quantifies this. It is equal to the larger of the following two numbers: (1) the difference between false-positive rates of the two groups, (2) the difference between false-negative rates of the two groups.\n",
+    "\n",
+    "The table above shows the overall AUC of 0.85 (based on continuous predictions) and the overall balance rate of 0.22 (based on 0/1 predictions). Both of these are satisfactory in our application context. However, there is a large disparity in accuracy rates (as indicated by the balanced error rate difference) and even larger when we consider the equalized-odds difference. As a sanity check, we also show the demographic parity ratio, whose level (slightly above 0.8) is considered satisfactory in this context."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Mitigating Equalized Odds Difference with Postprocessing"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We attempt to mitigate the disparities in the `lightgbm` prediction using the Fairlearn postprocessing algorithm `ThresholdOptimizer`. This algorithm finds a suitable threshold for the scores (class probabilities) produced by the `lightgbm` model by optimizing the accuracy rate under the constraint that the equalized odds difference (on training data) is zero. Since our goal is to optimize balanced accuracy, we resample the training data to have the same number of positive and negative examples. (This means that `ThresholdOptimizer` is effectively optimizing balanced accuracy on the original data.)"
```

Comment:
The last sentence in this paragraph is entirely in parentheses. This means that it shouldn't be.

(And git and notebooks are really not playing nicely here)

Reply:
fixed.